includes/class-wporg-gp-translation-events-translation-listener.php

```diff
@@ -50,71 +56,102 @@ function ( $translation, $translation_before ) {
+			/** @var GP_Translation_Set $translation_set */
+			$translation_set = ( new GP_Translation_Set )->find_one( [ 'id' => $translation->translation_set_id ] );
+			global $wpdb;
+			$table_name = self::ACTIONS_TABLE_NAME;
+
+			foreach ( $events as $event ) {
+				// A given user can only do one action on a specific translation.
+				// So we insert ignore, which will keep only the first action.
+				$wpdb->query(
+					$wpdb->prepare(
+						"insert ignore into $table_name (event_id, user_id, translation_id, action, locale) values (%d, %d, %d, %s, %s)",
+						[
+							// start primary key
+							'event_id'       => $event->id(),
+							'user_id'        => $user_id,
+							'translation_id' => $translation->id,
+							// end primary key
+							'action'         => $action,
+							'locale'         => $translation_set->locale,
+						],
+					),
+				);
+			}
+		} catch ( Exception $exception ) {
+			error_log( $exception );
 		}
 	}
 
 	/**
-	 * @return WP_Post[]
+	 * @return WPORG_GP_Translation_Events_Event[]
+	 * @throws Exception
 	 */
-	private function get_active_events( DateTime $at ): array {
-		return get_posts(
-			[
-				'post_type'   => 'event',
-				'post_status' => 'publish',
-				'meta_query'  => [
-					[
-						'key'     => '_event_start',
-						'value'   => $at->format( 'Y-m-d H:i:s' ),
-						'compare' => '<=',
-						'type'    => 'DATETIME',
-					],
-					[
-						'key'     => '_event_end',
-						'value'   => $at->format( 'Y-m-d H:i:s' ),
-						'compare' => '>=',
-						'type'    => 'DATETIME',
+	private function get_active_events( DateTimeImmutable $at ): array {
+		$events = $this->active_events_cache->get();
+		if ( null === $events ) {
+			$cache_duration = WPORG_GP_Translation_Events_Active_Events_Cache::CACHE_DURATION;
+			$boundary_start = $at;
+			$boundary_end   = $at->modify( "+$cache_duration seconds" );
+
+			// Get events for which start is before $boundary_end AND end is after $boundary_start.
+			$event_ids = get_posts(
+				[
+					'post_type'      => 'event',
+					'post_status'    => 'publish',
+					'posts_per_page' => - 1,
+					'fields'         => 'ids',
+					'meta_query'     => [
+						[
+							'key'     => '_event_start',
+							'value'   => $boundary_end->format( 'Y-m-d H:i:s' ),
+							'compare' => '<',
+							'type'    => 'DATETIME',
+						],
+						[
+							'key'     => '_event_end',
+							'value'   => $boundary_start->format( 'Y-m-d H:i:s' ),
+							'compare' => '>',
+							'type'    => 'DATETIME',
+						],
 					],
 				],
-			],
+			);
+
+			$events = [];
+			foreach ( $event_ids as $event_id ) {
+				$meta = get_post_meta( $event_id );
+				$events[] = WPORG_GP_Translation_Events_Event::fromPostMeta($event_id, $meta);
+			}
+
+			$this->active_events_cache->cache( $events );
+		}
+
+		// Filter out events that aren't actually active at $at.
+		return array_filter(
+			$events,
+			function ( $event ) use ( $at ) {
+				return $at >= $event->start() && $at <= $event->end();
```

Comment:
Just mentioning this as an alternative formatting which makes it a tad clearer that the $at is between the two events.
```suggestion
				return $event->start() <= $a && $at <= $event->end();
```